### PR TITLE
Add "Aim Assist" mod

### PR DESCRIPTION
**Pull request by MaxOhn**
Rough cut of the AimAssist mod, inspired by [this video](https://www.youtube.com/watch?v=afI_6X0PdM8).
HitCircles and Sliders move towards the cursor and sliders continue moving so that the SliderBall sticks to the cursor.
Including the overall way I implemented it which probably has a much nicer solution, there's still a couple of other things to fix.
Also references the issue #5760

Preview: https://streamable.com/srtful

FIXME:
- [x] Some DrawableHitCircles resulting in a flash at the original(?) position when clicked too early
- [x] Some DrawableSliders appearing at their original position for one frame when they're done
- [x] ~~Don't add spinner progress for cursor movement by user~~
- [x] ~~Handle spinners more consistently~~

### CODE REVIEW

File: osu.Game.Rulesets.Osu/Mods/OsuModAimAssist.cs [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+                            else
+                                d.MoveTo(drawableCursor.DrawPosition - slider.Ball.DrawPosition, Math.Max(0, h.StartTime - currentTime));
+                            return currentTime > (h as IHasEndTime)?.EndTime - 50;
+
+                        case DrawableSpinner _:
+                            // TODO
+                            return true;
+                    }
+
+                    return true; // never happens(?)
```
Comment (iiSaLMaN) [inline osu.Game.Rulesets.Osu/Mods/OsuModAimAssist.cs]: just set a default case, probably on `DrawableHitCircle`

File: osu.Game.Rulesets.Osu/Mods/OsuModAimAssist.cs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+                drawable.ApplyCustomUpdateState += drawableOnApplyCustomUpdateState;
+        }
+
+        private void drawableOnApplyCustomUpdateState(DrawableHitObject drawable, ArmedState state)
+        {
+            if (!(drawable is DrawableOsuHitObject d))
+                return;
+
+            var h = d.HitObject;
+            using (d.BeginAbsoluteSequence(h.StartTime - h.TimePreempt))
```
Comment (iiSaLMaN) [inline osu.Game.Rulesets.Osu/Mods/OsuModAimAssist.cs]: I don't see any transforms..?

File: osu.Game.Rulesets.Osu/Mods/OsuModAimAssist.cs [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        private readonly HashSet<DrawableOsuHitObject> movingObjects = new HashSet<DrawableOsuHitObject>();
+        private int updateCounter;
+
+        public void Update(Playfield playfield)
+        {
+            // Avoid crowded judgment displays
+            playfield.DisplayJudgements.Value = false;
+
+            // Object destination updated when cursor updates
+            playfield.Cursor.ActiveCursor.OnUpdate += drawableCursor =>
```
Comment (iiSaLMaN) [inline osu.Game.Rulesets.Osu/Mods/OsuModAimAssist.cs]: Subscribing to the cursor update every playfield update?? you probably want `IApplicableToDrawableRuleset` instead

File: osu.Game.Rulesets.Osu/Mods/OsuModAimAssist.cs [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+        public void ApplyToDrawableHitObjects(IEnumerable<DrawableHitObject> drawables)
+        {
+            foreach (var drawable in drawables)
+                drawable.ApplyCustomUpdateState += drawableOnApplyCustomUpdateState;
+        }
+
+        private void drawableOnApplyCustomUpdateState(DrawableHitObject drawable, ArmedState state)
+        {
+            if (!(drawable is DrawableOsuHitObject d))
```
Comment (iiSaLMaN) [inline osu.Game.Rulesets.Osu/Mods/OsuModAimAssist.cs]:
```suggestion
            if (drawable is DrawableOsuHitObject d)
                movingObjects.Add(d);
```

File: osu.Game.Rulesets.Osu/Mods/OsuModAimAssist.cs [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    {
+        public override string Name => "AimAssist";
+        public override string Acronym => "AA";
+        public override IconUsage Icon => FontAwesome.Solid.MousePointer;
+        public override ModType Type => ModType.Fun;
+        public override string Description => "No need to chase the circle, the circle chases you";
+        public override double ScoreMultiplier => 1;
+        public override Type[] IncompatibleMods => new[] { typeof(OsuModAutopilot), typeof(OsuModAutoplay), typeof(OsuModWiggle), typeof(OsuModTransform) };
+
+        private readonly HashSet<DrawableOsuHitObject> movingObjects = new HashSet<DrawableOsuHitObject>();
```
Comment (iiSaLMaN) [inline osu.Game.Rulesets.Osu/Mods/OsuModAimAssist.cs]: why is this a hashset...?

**Review by frenzibyte — CHANGES_REQUESTED** [review]

File: osu.Game.Rulesets.Osu/Mods/OsuModAimAssist.cs [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+                movingObjects.RemoveWhere(d =>
+                {
+                    var currentTime = playfield.Clock.CurrentTime;
+                    var h = d.HitObject;
+                    d.ClearTransforms();
+
+                    switch (d)
+                    {
+                        case DrawableHitCircle circle:
+                            circle.MoveTo(drawableCursor.DrawPosition, Math.Max(0, h.StartTime - currentTime));
```
Comment (iiSaLMaN) [inline osu.Game.Rulesets.Osu/Mods/OsuModAimAssist.cs]: Preempt time of the object is what you want
```suggestion
                            circle.MoveTo(drawableCursor.DrawPosition, h.TimePreempt);
```

**Comment by Desconocidosmh** [comment]
Did you think about making it snap to the cursor only at a certain range? I think it would make this more of a separate mod that makes the circles feel "bigger", therefore it would be more of an actual "assist", rather that being basically the autopilot in disguise.

**Comment by MaxOhn** [comment]
@Desconocidosmh Not uninteresting but I feel like there is little incentive to play such a mod, especially because it would be unranked anyway. I'm leaning more towards Aim Assist being a laid-back fun mod. It's an idea for a potential mod  in the future though 😄

File: osu.Game.Rulesets.Osu/Mods/OsuModAimAssist.cs [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+        }
+    }
+
+    /*
+     * TODOs
+     *  - fix sliders reappearing at original position after their EndTime (see https://puu.sh/E7zT4/111cf9cdc8.gif)
+     *  - find nicer way to handle slider headcircle explosion, flash, ...
+     *  - add Aim Assist as incompatible mod for Autoplay (?)
+     *
+     */
```
Comment (HoLLy-HaCKeR) [inline osu.Game.Rulesets.Osu/Mods/OsuModAimAssist.cs]: You probably want to place this in your issue, so it is visible for everyone.

**Comment by HoLLy-HaCKeR** [comment]
Should this be a Multimod with Relax?

File: osu.Game.Rulesets.Osu/Mods/OsuModAimAssist.cs [hunk 8]
```diff
@@ -0,0 +1,137 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
```
Comment (WheelChairUser420) [inline osu.Game.Rulesets.Osu/Mods/OsuModAimAssist.cs]: Nerf shit

**Comment by peppy** [comment]
@MaxOhn shall we close this one for now? seems the TODO items you had are still present.

**Comment by MaxOhn** [comment]
Oh right I put this project aside and completely forgot about it, my apologies.
I'll see to get into it again in the next few days, otherwise I'll close this within the next week-ish.

**Comment by MaxOhn** [comment]
I've updated everything w.r.t. the current lazer state to the best of my abilities.
Unfortunately, I'm stuck on the remaining issues I listed in the top comment and I'm honestly not too keen on spending more time to figuring them out.
Unless there is interest from others, I'd be fine with closing this.

**Comment by peppy** [comment]
Thanks for bringing it up-to-date. I'll try and fit in a review pass and see if the issues are simple enough to fix (some sound like they should be quick fixes).

**Comment by peppy** [comment]
Actually not sure how simple this will be to get right. For a start, you should be using interpolation rather than `MoveTo`, or at least rate limiting the usage of `MoveTo` to avoid transforming every frame. Maybe the latter as it's the only want to make rewind semi-consistent.

**Comment by MaxOhn** [comment]
Assist strength is now configurable. I don't think spinners and ongoing sliders should be modified on partial assistance so I now ignore them.
I'm not sure how to ratelimit the update properly. For now I just update once every 100ms but I'm assuming that could cause inconsistencies.
The notes flashing in their original position is still an issue I'm also clueless about.

**Comment by peppy** [comment]
I've done an initial refactor and clean-up pass on this after discussion with @smoogipoo. It should be in a visually correct state now, but replay support is ad-hoc at best. Making it work well in replays is potentially achievable, but will add a lot of complexity that I'd rather avoid in an initial implementation.

Note that I brought back the removed slider behaviour (tracking the sliders) as it feels really good, but I have not added the spinner spinning. I think spinner support could work if done well, but tweaking it will require some dedication, of which I don't have right now.

**Review by peppy — DISMISSED** [review]

**Comment by bdach** [comment]
I've pushed 7463744 to fix a minor issue - `OsuModAimAssist` was declaring incompatibility with `ModAutoplay`, but the osu!-specific autoplay implementations weren't declaring incompatibility with `OsuModAimAssist`, resulting in some weird behaviours.

Before:

https://user-images.githubusercontent.com/20418176/152217209-de74ed82-922c-41cc-a4df-de3d3f6e266f.mp4

After:

https://user-images.githubusercontent.com/20418176/152217197-3eb2f479-0ba8-4fe2-9d21-29b47516c847.mp4

Note the aim assist mod missing from tooltips on first video, as well as the weird behaviour of the autoplay button (blocked from operation when aim assist is selected).

Other than that this seems okay. I will say that as someone who likes to turn everything off (I turn off all HUD entirely in gameplay, as I don't really ever look at the HP bar or the hit error meters), turning judgements off is a bit of an issue as I don't get any timing feedback anymore. That said I don't really feel like spending time on fixing that for this mod, so possibly fine. I'll leave it to someone else to merge.

**Review by bdach — APPROVED** [review]

**Comment by peppy** [comment]
> Other than that this seems okay. I will say that as someone who likes to turn everything off (I turn off all HUD entirely in gameplay, as I don't really ever look at the HP bar or the hit error meters), turning judgements off is a bit of an issue as I don't get any timing feedback anymore. That said I don't really feel like spending time on fixing that for this mod, so possibly fine. I'll leave it to someone else to merge.

I did have a quick look at what would be required to fix judgements, but unless doing a mod level hack to update their positions it does seem non trivial due to the data flow.

**Pull request merged**